modules/joinmarket.nix

```diff
@@ -141,75 +141,78 @@ in {
-        description = "joinmarket User";
-        group = "${cfg.group}";
-        home = cfg.dataDir;
-        extraGroups = [ "tor" ];
-    };
-    users.groups.${cfg.group} = {};
-    nix-bitcoin.operator = {
-      groups = [ cfg.group ];
-      sudoUsers = [ cfg.group ];
-    };
 
     systemd.tmpfiles.rules = [
       "d '${cfg.dataDir}' 0770 ${cfg.user} ${cfg.group} - -"
     ];
 
-    services.bitcoind.disablewallet = false;
-
-    # Joinmarket is TOR-only
-    services.tor = {
-      enable = true;
-      client.enable = true;
-      controlSocket.enable = true;
-    };
-
     systemd.services.joinmarket = {
-      description = "JoinMarket Daemon";
       wantedBy = [ "multi-user.target" ];
       requires = [ "bitcoind.service" ];
       after = [ "bitcoind.service" ];
       path = [ pkgs.sudo ];
       serviceConfig = nix-bitcoin-services.defaultHardening // {
-        ExecStartPre = nix-bitcoin-services.privileged ''
+        ExecStartPre = nix-bitcoin-services.privileged "joinmarket-create-config" ''
           install -o '${cfg.user}' -g '${cfg.group}' -m 640 ${configFile} ${cfg.dataDir}/joinmarket.cfg
           sed -i \
              "s|@@RPC_PASSWORD@@|rpc_password = $(cat ${secretsDir}/bitcoin-rpcpassword-privileged)|" \
              '${cfg.dataDir}/joinmarket.cfg'
         '';
         # Generating wallets (jmclient/wallet.py) is only supported for mainnet or testnet
-        ExecStartPost = mkIf (bitcoind.network == "mainnet") (nix-bitcoin-services.privileged ''
-          walletname=wallet.jmdat
-          pw=$(cat "${secretsDir}"/jm-wallet-password)
-          mnemonic=${secretsDir}/jm-wallet-seed
-          if [[ ! -f ${cfg.dataDir}/wallets/$walletname ]]; then
-            echo Create joinmarket wallet
-            # Use bash variables so commands don't proceed on previous failures
-            # (like with pipes)
-            cd ${cfg.dataDir} && \
-              out=$(sudo -u ${cfg.user} \
-              ${nbPkgs.joinmarket}/bin/jm-genwallet \
-              --datadir=${cfg.dataDir} $walletname $pw)
-            recoveryseed=$(echo "$out" | grep 'recovery_seed')
-            echo "$recoveryseed" | cut -d ':' -f2 > $mnemonic
-          fi
-        '');
+        ExecStartPost = mkIf (bitcoind.network == "mainnet")
+          (nix-bitcoin-services.privileged "joinmarket-create-wallet" ''
+            walletname=wallet.jmdat
+            if [[ ! -f ${cfg.dataDir}/wallets/$walletname ]]; then
+              echo "Create wallet"
+              pw=$(cat "${secretsDir}"/jm-wallet-password)
+              cd ${cfg.dataDir}
+              # Instead of redirecting the pipe output to the seed file, store it in a variable.
+              # This prevents writing an empty seed file if the pipe fails
+              walletSeed=$(
+                sudo -u ${cfg.user} ${nbPkgs.joinmarket}/bin/jm-genwallet --datadir=${cfg.dataDir} $walletname $pw \
+                  | grep 'recovery_seed' \
+                  | cut -d ':' -f2
```

Comment:
I tried again, this time replacing with `cut -d '--' -f2` and the script again succeeded without producing a seed. Can we add `set -u` to make extra sure this script doesn't fail?

Reply:
The initial version of my refactored script behaved in exactly the same way as your version (including failures), except
that my script wrote an empty seed file when the pipe failed (this was fixed by the last squash).
The reason that you saw joinmarket running on your node after a pipe failure is that joinmarket was automatically restarted.
But on the second run it skipped the wallet generation because the wallet file from the failed previous run still existed.
The same thing happened with your original script.
[The fix is](https://github.com/erikarvstedt/nix-bitcoin/commit/6e7b951) to remove the wallet when seed extraction fails.